File: src/Asset/Vns.Erp.Core.Asset.CsAssetGUI/DanhMuc/frmDanhMucLoaiTaiSan1_ChiTiet.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;

using Vns.Erp.Core;
using Vns.Erp.Core.Asset.Domain;
using Vns.Erp.Core.Asset.Service.Interface;

namespace Vns.Erp.Core.Asset.CsAssetGUI
{
    public partial class frmDanhMucLoaiTaiSan1_ChiTiet : DevExpress.XtraEditors.XtraForm
    {

        #region "Variables and Messages"
        private FormGlobals.DataInputState m_InputState;
        private Guid m_ID;
        private DmLoaitaisan objInfo;
        private string p_ma;
        IDmLoaitaisanService _DmLoaiTaisanService;
        public Guid ID
        {
            get { return m_ID; }
            set { m_ID = value; }
        }

        public virtual IDmLoaitaisanService DmLoaiTaisanService
        {
            get { return _DmLoaiTaisanService; }
            set { _DmLoaiTaisanService = value; }
        }
        #endregion

        #region "Load Form"

        public frmDanhMucLoaiTaiSan1_ChiTiet()
        {
            InitializeComponent();
            KeyDown += frmDanhMucTuDo1_ChiTiet_KeyDown;
            TNCommon.setIconControl(this);
        }

        public Guid Show_Form(string iID, FormGlobals.DataInputState eState)
        {
            try
            {
                FormGlobals.Form_SetText(this, "Danh mục loại tài sản chi tiết", FormGlobals.DataInputState.NoCaption);
                FormGlobals.Form_InitDialog(this);

                m_InputState = eState;
                if ((eState.Equals(FormGlobals.DataInputState.EditMode)))
                {
                    m_ID = new Guid(iID);
                }
                else
                {
                    m_ID = Guid.NewGuid();
                }

                frmProgress.Instance().Thread = Init_Form;
                frmProgress.Instance().Show_Progress();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }

            if ((this.ShowDialog() == DialogResult.OK))
            {
                return m_ID;
            }
            else
            {
                return Null.NullGuid;
            }
        }

        #endregion

        #region "Private Function and Procedures"

        private void Init_Control()
        {
            FormGlobals.Control_SetRequire(txtKy_Hieu, true);
            FormGlobals.Control_SetRequire(cboLOAITAISAN_ID_CHA, true);
            FormGlobals.Control_SetRequire(txtTEN_LOAITAISAN, true);
            FormGlobals.Control_SetFont(this, FormGlobals.CS_FONT_NAME);
            //Panel_InitControl(Me._GroupControlObjectInfo, "DM_LOAITAISAN")
        }

        private void Init_Form()
        {
            Init_Control();
            Init_Data();
        }
        private void ClearForm()
        {
            this.txtKHUNG_TGSD1.Text = "";
            this.txtKy_Hieu.Text = "";
            this.txtTEN_LOAITAISAN.Text = "";
        }
        private void Init_Data()
        {
            switch (m_InputState)
            {
                case FormGlobals.DataInputState.AddMode:
                    BindLookUpEdit();
                    ClearForm();
                    TNCommon.SelectFirst(cboLOAITAISAN_ID_CHA);
                    break;
                case FormGlobals.DataInputState.EditMode:
                    BindLookUpEdit();
                    Load_Data();
                    break;
            }
        }

        private void Load_Data()
        {
            objInfo = DmLoaiTaisanService.GetById(m_ID);
            p_ma = objInfo.KyHieu;
            this.txtKy_Hieu.Text = objInfo.KyHieu;
            this.txtTEN_LOAITAISAN.Text = objInfo.TenLoaitaisan;
            this.txtKHUNG_TGSD1.Text = objInfo.KhungTgsd.ToString();
            this.cboLOAITAISAN_ID_CHA.EditValue = objInfo.LoaitaisanIdCha.ToString();
            //Panel_SetControlValue(Me._GroupControlObjectInfo, objInfo, True)
            txtKHUNG_TGSD1.Text = objInfo.KhungTgsd.ToString();
            if (objInfo.LoaitaisanIdCha == null)
            {
                cboLOAITAISAN_ID_CHA.EditValue = null;
            }
            else
            {
                cboLOAITAISAN_ID_CHA.EditValue = objInfo.LoaitaisanIdCha;
            }

            chkCO_SU_DUNG.Checked = TNCommon.ParseBool(objInfo.CoSuDung);
        }

        private void Save_Data()
        {
            if (objInfo == null)
            {
                objInfo = new DmLoaitaisan();
            }
            objInfo.KyHieu = this.txtKy_Hieu.Text;
            objInfo.TenLoaitaisan = txtTEN_LOAITAISAN.Text;
            objInfo.CoSuDung = TNCommon.ParseDecimal(chkCO_SU_DUNG.Checked);
            objInfo.LoaitaisanIdCha = VnsConvert.CGuid(cboLOAITAISAN_ID_CHA.EditValue);
            if (!string.IsNullOrEmpty(txtKHUNG_TGSD1.Text))
            {
                objInfo.KhungTgsd = decimal.Parse(txtKHUNG_TGSD1.Text);
            }

            //Try
            if (m_InputState == FormGlobals.DataInputState.AddMode || m_InputState == FormGlobals.DataInputState.CopyMode)
            {
                objInfo.MaLoaitaisan = " ";
                objInfo.Id = new Guid();
                DmLoaiTaisanService.SaveOrUpdate(objInfo);
                m_ID = objInfo.Id;
            }
            else
            {
                DmLoaiTaisanService.SaveOrUpdate(objInfo);
                m_ID = objInfo.Id;
            }
            frmProgress.Instance().SetFinishText("Data saved");
            //Catch ex As Exception
            //    Message_Error(ex)
            //End Try

        }

        private void BindLookUpEdit()
        {
            //Bind danh muc cha
            DmLoaitaisan dr = default(DmLoaitaisan);
            List<DmLoaitaisan> dt = new List<DmLoaitaisan>();
            dt.AddRange(DmLoaiTaisanService.GetAll());
            dr = new DmLoaitaisan();
            dr.TenLoaitaisan = "Danh mục gốc";
            dr.Id = Guid.Empty;
            dr.MaLoaitaisan = "";
            dr.Cap = 1;
            dt.Insert(0, dr);


            cboLOAITAISAN_ID_CHA.Properties.ValueMember = "Id";
            cboLOAITAISAN_ID_CHA.Properties.DisplayMember = "TenLoaitaisan";
            cboLOAITAISAN_ID_CHA.Properties.DataSource = dt;
        }

        private bool Check_KH(string KH)
        {
            List<DmLoaitaisan> dt = new List<DmLoaitaisan>();
            dt.AddRange(DmLoaiTaisanService.GetByKyHieu(KH));
            if (dt.Count != 0)
            {
                FormGlobals.Message_Warning("Ký hiệu loại tài sản đã tồn tại");
                txtKy_Hieu.Focus();
                return false;
            }
            return true;
        }

        private bool Check_input()
        {
            if (string.IsNullOrEmpty(txtKy_Hieu.Text) ||
                string.IsNullOrEmpty(txtTEN_LOAITAISAN.Text))
            {
                FormGlobals.Message_Warning("Bạn chưa nhập giá trị cho các trường bắt buộc");
                return false;
            }

            if (!string.IsNullOrEmpty(txtKHUNG_TGSD1.Text))
            {
                decimal Nam = decimal.Parse(txtKHUNG_TGSD1.Text);
                if (Nam < 0 | Nam > 9999)
                {
                    FormGlobals.Message_Warning("Số Năm sử dụng không hợp lệ!");
                    txtKHUNG_TGSD1.Focus();
                    return false;
                }
            }

            //if (VnsCheck.IsNullGuid(cboLOAITAISAN_ID_CHA.EditValue))
            //{
            //    FormGlobals.Message_Warning("Bạn chưa chọn danh mục cha!");
            //    cboLOAITAISAN_ID_CHA.Focus();
            //    return false;
            //}

            return true;
        }
        #endregion

        #region "Events"

        private void btnCancel_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                this.Close();
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnClear_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                FormGlobals.Control_ClearData(_GroupControlObjectInfo);
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void btnSave_Click(System.Object sender, System.EventArgs e)
        {
            try
            {
                string kh = "";
                switch (m_InputState)
                {
                    case FormGlobals.DataInputState.AddMode:
                        kh = txtKy_Hieu.Text;
                        break;
                    case FormGlobals.DataInputState.EditMode:
                        if (!p_ma.Equals(txtKy_Hieu.Text))
                        {
                            kh = txtKy_Hieu.Text;
                        }
                        else
                        {
                            kh = "";
                        }
                        break;
                }

                if (!Check_KH(kh))
                {
                    return;
                }

                if (Check_input())
                {
                    frmProgress.Instance().Thread = Save_Data;
                    frmProgress.Instance().Show_Progress("Saving data");
                    DialogResult = System.Windows.Forms.DialogResult.OK;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        private void frmDanhMucTuDo1_ChiTiet_KeyDown(System.Object sender, System.Windows.Forms.KeyEventArgs e)
        {
            try
            {
                switch (e.KeyCode)
                {
                    case Keys.Escape:
                        this.Close();
                        break; // TODO: might not be correct. Was : Exit Select

                        break;
                }
            }
            catch (Exception ex)
            {
                FormGlobals.Message_Error(ex);
            }
        }

        #endregion

        private void btnClear_Click_1(object sender, EventArgs e)
        {

        }



    }

}

File: src/Admin/Vns.Erp.Core.Admin.VbAdminGUI/DanhMuc/frmDanhMucLoaiTaiSan1_ChiTiet.vb
Imports System
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Collections
Imports Vns.Erp.Core.Admin.Domain
Imports Vns.Erp.Core.Admin.Service.Interface

Public Class frmDanhMucLoaiTaiSan1_ChiTiet

#Region "Variables and Messages"
    Private m_InputState As FormGlobals.DataInputState
    Private m_ID As Guid
    Private objInfo As DmLoaitaisan
    Private p_ma As String
    Dim _DmLoaiTaisanService As IDmLoaitaisanService
    Public Property ID() As Guid
        Get
            Return m_ID
        End Get
        Set(value As Guid)
            m_ID = value
        End Set
    End Property

    Public Overridable Property DmLoaiTaisanService() As IDmLoaitaisanService
        Get
            Return _DmLoaiTaisanService
        End Get
        Set(ByVal value As IDmLoaitaisanService)
            _DmLoaiTaisanService = value
        End Set
    End Property
#End Region

#Region "Load Form"

    Public Sub New()
        InitializeComponent()
        TNCommon.setIconControl(Me)
    End Sub

    Public Function Show_Form(ByVal iID As String, ByVal eState As DataInputState) As Guid
        Try
            Form_SetText(Me, "Danh mục loại tài sản chi tiết", DataInputState.NoCaption)
            Form_InitDialog(Me)

            m_InputState = eState
            If (eState.Equals(DataInputState.EditMode)) Then
                m_ID = New Guid(iID)
            Else
                m_ID = Guid.NewGuid()
            End If

            frmProgress.Instance.Thread = AddressOf Init_Form
            frmProgress.Instance.Show_Progress()
        Catch ex As Exception
            Message_Error(ex)
        End Try

        If (Me.ShowDialog = DialogResult.OK) Then
            Return m_ID
        Else
            Return Nothing
        End If
    End Function

#End Region

#Region "Private Function and Procedures"

    Private Sub Init_Control()
        Control_SetRequire(txtKy_Hieu, True)
        Control_SetRequire(cboLOAITAISAN_ID_CHA, True)
        Control_SetRequire(txtTEN_LOAITAISAN, True)
        Control_SetFont(Me, CS_FONT_NAME)
        'Panel_InitControl(Me._GroupControlObjectInfo, "DM_LOAITAISAN")
    End Sub

    Private Sub Init_Form()
        Init_Control()
        Init_Data()
    End Sub
    Private Sub ClearForm()
        Me.txtKHUNG_TGSD1.Text = ""
        Me.txtKy_Hieu.Text = ""
        Me.txtTEN_LOAITAISAN.Text = ""
    End Sub
    Private Sub Init_Data()
        Select Case m_InputState
            Case DataInputState.AddMode
                BindLookUpEdit()
                ClearForm()
                TNCommon.SelectFirst(cboLOAITAISAN_ID_CHA)
            Case DataInputState.EditMode
                BindLookUpEdit()
                Load_Data()
        End Select
    End Sub

    Private Sub Load_Data()
        objInfo = DmLoaiTaisanService.GetById(m_ID)
        p_ma = objInfo.KyHieu
        Me.txtKy_Hieu.Text = objInfo.KyHieu
        Me.txtTEN_LOAITAISAN.Text = objInfo.TenLoaitaisan
        Me.txtKHUNG_TGSD1.Text = objInfo.KhungTgsd.ToString
        Me.cboLOAITAISAN_ID_CHA.EditValue = objInfo.LoaitaisanIdCha.ToString
        'Panel_SetControlValue(Me._GroupControlObjectInfo, objInfo, True)
        txtKHUNG_TGSD1.Text = objInfo.KhungTgsd.ToString
        If objInfo.LoaitaisanIdCha = Nothing Then
            cboLOAITAISAN_ID_CHA.EditValue = Nothing
        Else
            cboLOAITAISAN_ID_CHA.EditValue = objInfo.LoaitaisanIdCha
        End If

        chkCO_SU_DUNG.Checked = TNCommon.ParseBool(objInfo.CoSuDung)
    End Sub

    Private Sub Save_Data()
        If objInfo Is Nothing Then
            objInfo = New DmLoaitaisan
        End If
        objInfo.KyHieu = Me.txtKy_Hieu.Text
        objInfo.TenLoaitaisan = txtTEN_LOAITAISAN.Text
        objInfo.CoSuDung = TNCommon.ParseDecimal(chkCO_SU_DUNG.Checked)
        objInfo.LoaitaisanIdCha = New Guid(cboLOAITAISAN_ID_CHA.EditValue.ToString())
        If txtKHUNG_TGSD1.Text <> "" Then
            objInfo.KhungTgsd = Decimal.Parse(txtKHUNG_TGSD1.Text)
        End If

        'Try
        If m_InputState = DataInputState.AddMode OrElse m_InputState = DataInputState.CopyMode Then
            objInfo.MaLoaitaisan = " "
            objInfo.Id = New Guid()
            DmLoaiTaisanService.SaveOrUpdate(objInfo)
            m_ID = objInfo.Id
        Else
            DmLoaiTaisanService.SaveOrUpdate(objInfo)
            m_ID = objInfo.Id
        End If
        frmProgress.Instance.SetFinishText("Data saved")
        'Catch ex As Exception
        '    Message_Error(ex)
        'End Try

    End Sub

    Private Sub BindLookUpEdit()
        'Bind danh muc cha
        Dim dr As DmLoaitaisan
        Dim dt As List(Of DmLoaitaisan) = New List(Of DmLoaitaisan)
        dt.AddRange(DmLoaiTaisanService.GetAll())
        dr = New DmLoaitaisan()
        dr.TenLoaitaisan = "Danh mục gốc"
        dr.Id = Guid.Empty
        dr.MaLoaitaisan = ""
        dr.Cap = 1
        dt.Insert(0, dr)


        cboLOAITAISAN_ID_CHA.Properties.ValueMember = "Id"
        cboLOAITAISAN_ID_CHA.Properties.DisplayMember = "TenLoaitaisan"
        cboLOAITAISAN_ID_CHA.Properties.DataSource = dt
    End Sub

    Private Function Check_KH(ByVal KH As String) As Boolean
        Dim dt As List(Of DmLoaitaisan) = New List(Of DmLoaitaisan)
        dt.AddRange(DmLoaiTaisanService.GetByKyHieu(KH))
        If dt.Count <> 0 Then
            Message_Warning("Ký hiệu loại tài sản đã tồn tại")
            txtKy_Hieu.Focus()
            Return False
        End If
        Return True
    End Function

    Private Function Check_input() As Boolean
        If (txtKy_Hieu.Text = "" Or txtTEN_LOAITAISAN.Text = "" Or DB_GetString(cboLOAITAISAN_ID_CHA.EditValue) = "") Then
            Message_Warning("Bạn chưa nhập giá trị cho các trường bắt buộc")
            Return False
        End If

        If txtKHUNG_TGSD1.Text <> "" Then
            Dim Nam As Decimal = Decimal.Parse(txtKHUNG_TGSD1.Text)
            If Nam < 0 Or Nam > 9999.99 Then
                Message_Warning("Số Năm sử dụng không hợp lệ!")
                txtKHUNG_TGSD1.Focus()
                Return False
            End If
        End If

        If DB_GetString(cboLOAITAISAN_ID_CHA.EditValue) = "" Then
            Message_Warning("Bạn chưa chọn danh mục cha!")
            cboLOAITAISAN_ID_CHA.Focus()
            Return False
        End If

        Return True
    End Function
#End Region

#Region "Events"

    Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click
        Try
            Me.Close()
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnClear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClear.Click
        Try
            Control_ClearData(_GroupControlObjectInfo)
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub btnSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSave.Click
        Try
            Dim kh As String = ""
            Select Case m_InputState
                Case DataInputState.AddMode
                    kh = txtKy_Hieu.Text
                Case DataInputState.EditMode
                    If Not p_ma.Equals(txtKy_Hieu.Text) Then
                        kh = txtKy_Hieu.Text
                    Else
                        kh = ""
                    End If
            End Select

            If Not Check_KH(kh) Then
                Exit Sub
            End If

            If Check_input() Then
                frmProgress.Instance.Thread = AddressOf Save_Data
                frmProgress.Instance.Show_Progress("Saving data")
                DialogResult = Windows.Forms.DialogResult.OK
            End If
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

    Private Sub frmDanhMucTuDo1_ChiTiet_KeyDown(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyDown
        Try
            Select Case e.KeyCode
                Case Keys.Escape
                    Me.Close()
                    Exit Select
            End Select
        Catch ex As Exception
            Message_Error(ex)
        End Try
    End Sub

#End Region

End Class

